src/tangent_types/thunks.jl

```diff
@@ -59,12 +59,41 @@ LinearAlgebra.UpperTriangular(a::AbstractThunk) = UpperTriangular(unthunk(a))
 LinearAlgebra.Symmetric(a::AbstractThunk, uplo=:U) = Symmetric(unthunk(a), uplo)
 LinearAlgebra.Hermitian(a::AbstractThunk, uplo=:U) = Hermitian(unthunk(a), uplo)
 
-function LinearAlgebra.diagm(kv::Pair{<:Integer,<:AbstractThunk}...)
-    return diagm((k => unthunk(v) for (k, v) in kv)...)
-end
-function LinearAlgebra.diagm(m, n, kv::Pair{<:Integer,<:AbstractThunk}...)
-    return diagm(m, n, (k => unthunk(v) for (k, v) in kv)...)
+# diagm: not we only dispatch if occurs in first three positions do a few positions,
+# since we will incur type-piracy if we were to  unthunk any mix of thunk and nonthunked 
```

Comment:
Why would unthunking a mix be type piracy?   As soon as you require at least one `AbstractThunk` argument, it's no longer piracy, no?

Reply:
The `unthunk`ing step is fine.

It is the dispatch for `diagm` before hand.

For that we want 0 or more Pairs of any kind, then 1 Pair with a thunk, then 0 or more Pairs of any kind.
But Julia doesn't let us express the first part of that.

Reply:
I see, thanks.